src/typeChecker/typeErrorChecker.ts

```diff
@@ -251,7 +257,61 @@ function typeCheckAndReturnType(node: tsEs.Node, context: Context, env: TypeEnvi
       return tUndef
     }
     case 'CallExpression': {
-      const calleeType = typeCheckAndReturnType(node.callee, context, env)
+      const callee = node.callee
+      const args = node.arguments
+      if (context.chapter >= 2 && callee.type === 'Identifier') {
+        // Special functions for Source 2+: pair, list, head, tail
+        const fnName = callee.name
+        if (fnName === 'pair') {
+          if (args.length !== 2) {
+            context.errors.push(new InvalidNumberOfArgumentsTypeError(node, 2, args.length))
+            return tPair(tAny, tAny)
+          }
+          return tPair(
+            typeCheckAndReturnType(args[0], context, env),
+            typeCheckAndReturnType(args[1], context, env)
+          )
+        }
+        if (fnName === 'list') {
+          if (args.length === 0) {
+            return tNull
+          }
+          // Element type is union of all types of arguments in list
+          let elementType = typeCheckAndReturnType(args[0], context, env)
+          for (let i = 1; i < args.length; i++) {
+            elementType = mergeTypes(elementType, typeCheckAndReturnType(args[i], context, env))
+          }
+          // Type the list as a pair, for use when checking for type mismatches against pairs
+          let pairType = tPair(typeCheckAndReturnType(args[args.length - 1], context, env), tNull)
+          for (let i = args.length - 2; i >= 0; i--) {
+            pairType = tPair(typeCheckAndReturnType(args[i], context, env), pairType)
+          }
+          return tList(elementType, pairType)
+        }
+        if (fnName === 'head' || fnName === 'tail') {
+          if (args.length !== 1) {
+            context.errors.push(new InvalidNumberOfArgumentsTypeError(node, 1, args.length))
+            return tAny
+          }
+          const actualType = typeCheckAndReturnType(args[0], context, env)
+          // Argument should be either a pair or a list
+          const expectedType = tUnion(tPair(tAny, tAny), tList(tAny))
+          const numErrors = context.errors.length
+          checkForTypeMismatch(node, actualType, expectedType, context)
+          if (context.errors.length > numErrors) {
+            // If errors were found, return any type
+            return tAny
+          }
+          if (actualType.kind === 'pair') {
+            return fnName === 'head' ? actualType.headType : actualType.tailType
+          }
+          if (actualType.kind === 'list') {
+            return fnName === 'head' ? actualType.elementType : actualType
```

Comment:
Given that `List` is defined as such:
```ts
export interface List {
  kind: 'list'
  elementType: Type
  // Used in Source Typed variants to check for type mismatches against pairs
  typeAsPair?: Pair
}
```
If we take the `tail` of a list, the `typeAsPair` remains unchanged despite the list itself changing. Is this intended?

Reply:
Would also be good to add a test case that can catch this issue as a form of regression testing.

Reply:
Thanks for the catch! Will make the necessary changes.